core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1000,6 +1055,55 @@ class ReplicaManager(val config: KafkaConfig,
     result
   }
 
+  /**
+    * Using the configured [[ReplicaSelector]], determine the preferred read replica for a partition given the
+    * client metadata, the requested offset, and the current set of replicas. If the preferred read replica is the
+    * leader, return None
+    */
+  def findPreferredReadReplica(tp: TopicPartition,
+                               clientMetadata: ClientMetadata,
+                               replicaId: Int,
+                               fetchOffset: Long): Option[Int] = {
+    val partition = getPartitionOrException(tp, expectLeader = false)
+
+    if (partition.isLeader) {
+      if (Request.isValidBrokerId(replicaId)) {
+        // Don't look up preferred for follower fetches via normal replication
+        Option.empty
+      } else {
+        val replicaEndpoints = metadataCache.getPartitionReplicaEndpoints(tp.topic(), tp.partition(), new ListenerName(clientMetadata.listenerName))
+        val now = time.milliseconds
+        var replicaInfoSet: Set[ReplicaView] = partition.remoteReplicas
+          // Exclude replicas that don't have the requested offset (whether or not if they're in the ISR)
+          .filter(replica => replica.logEndOffset >= fetchOffset)
+          .filter(replica => replica.logStartOffset <= fetchOffset)
+          .map(replica => new DefaultReplicaView(
+            replicaEndpoints.getOrElse(replica.brokerId, Node.noNode()),
+            replica.logEndOffset,
+            now - replica.lastCaughtUpTimeMs
+          ))
+
+        if (partition.leaderReplicaIdOpt.isDefined) {
+          val leaderReplica: ReplicaView = partition.leaderReplicaIdOpt
+            .map(replicaId => replicaEndpoints.getOrElse(replicaId, Node.noNode()))
+            .map(leaderNode => new DefaultReplicaView(leaderNode, partition.localLogOrException.logEndOffset, 0L))
+            .get
+          replicaInfoSet ++= Set(leaderReplica)
+
+          val partitionInfo = new DefaultPartitionView(replicaInfoSet.asJava, leaderReplica)
+          replicaSelector.select(tp, clientMetadata, partitionInfo).asScala
+            .filter(!_.endpoint.isEmpty)
+            .filter(!_.equals(leaderReplica))
```

Comment:
nit: should be able to use `!=`

I think it's worth a comment explaining why we only return the preferred read replica if it is not the leader.